File: 1475-final-prices-with-a-special-discount-in-a-shop/1475-final-prices-with-a-special-discount-in-a-shop.java
import java.util.*;

class Solution {
    public int[] finalPrices(int[] prices) {
        int n = prices.length;
        Stack<Integer> stack = new Stack<>();
        for (int i = n - 1; i >= 0; i--) {
            while (!stack.isEmpty() && stack.peek() > prices[i]) {
                stack.pop();
            }
            int tmp = prices[i];
            if (!stack.isEmpty()) {
                prices[i] -= stack.peek();
            }
            stack.push(tmp);
        }
        return prices;
    }
}

File: 1475-final-prices-with-a-special-discount-in-a-shop/1475-final-prices-with-a-special-discount-in-a-shop.cpp
class Solution {
public:
    vector<int> finalPrices(vector<int>& prices) {
        int n = prices.size();
        stack<int>stk;
        for(int i = n-1 ; i >= 0 ; i--){
            while(stk.size() && stk.top() > prices[i])stk.pop();
            int tmp = prices[i];
            if(stk.size())prices[i]-=stk.top();
            stk.push(tmp);
        }
        return prices;
    }
};